Dev pathloss

brunohcfaria opened the pull request:
<h2>Result of the Sprint26</h2>
Now the simulator can import Base Station physical information from an input file as well as the path loss matrix.
The path loss information is implemented as "new" propagation model in the simulation structure.

brunohcfaria committed:
First version for the site cell data file for the Topology construction

brunohcfaria committed:
Added MacOS and PyCharm config files to gitignore

CalilQ committed:
Implement parsing of path loss files

Parsing is made by PropagationInputFile class

brunohcfaria committed:
* Implemented the parser for Base Station physical data file
* Created a parameter with BS information
* Added a sample physical data file - brucuCCO2600.xlsx

CalilQ committed:
Implement path loss heat map plot

Plot is run by the script at the end of propagation_input_file.py

brunohcfaria committed:
Merge branch 'dev-pathloss' of https://github.com/INDTSON/SHARC into dev-pathloss

CalilQ committed:
Implement get_loss method

Method receives array of cell IDs, UE x and y positions and returns path loss between BSs and UEs

CalilQ committed:
Merge branch 'dev-pathloss' of https://github.com/INDTSON/SHARC into dev-pathloss

CalilQ committed:
Improve Path Loss parser

Parser checks presence of necessary parameters in the header

brunohcfaria committed:
Added celdata input file information to create the topology and
the stations.

brunohcfaria committed:
Merge branch 'dev-pathloss' of https://github.com/INDTSON/SHARC into dev-pathloss

CalilQ committed:
Implement

CalilQ committed:
Merge branch 'dev-pathloss' of https://github.com/INDTSON/SHARC into dev-pathloss

CalilQ committed:
Pass correct get_loss parameters

Pass station_id instead of cell_id

CalilQ committed:
Fix bugs regarding propagation_input_map

brunohcfaria committed:
Refactored topology_input_map plot() method

CalilQ committed:
Fix minor simulation bugs

Includes cell definition for TopologyInputMap, as well as indoor stations and elevation as an array

brunohcfaria committed:
Merge branch 'dev-pathloss' of https://github.com/INDTSON/SHARC into dev-pathloss

brunohcfaria committed:
Added azimuth and elevation parameters from input file to TopolgyInputMap

CalilQ committed:
Fix major simulation bugs

Simulator now runs without crashing. Consistency of results is not guaranteed though.

CalilQ committed:
Switch X and Y coordinates in PropagationInputFile

Class was switching X and Y coordinates. Unit tests were updated.

brunohcfaria committed:
Added unit test for TopologyInputMap class

brunohcfaria committed:
Added unit test for input file parameter handling in ParametersImt

CalilQ committed:
Fix path loss sign

SHARC's convention is that the path loss is a positive number, while the input files list it as a negative number.

CalilQ committed:
Merge branch 'dev-pathloss' of https://github.com/INDTSON/SHARC into dev-pathloss

CalilQ committed:
Implement initial input file simulation tests

Test coverage and broadness needs to be improved though. Simulation still yields inconsistent results despite tests.

brunohcfaria committed:
minor fix in path loss file parsing in propagation_input_file.py

brunohcfaria commented on sharc/propagation/propagation_input_file.py:
```diff
@@ -35,28 +36,76 @@ def __init__(self, input_folder:str):
+                        else:
+                            idx_list.append(-4)
+                    
+                # Get parameter values
```
The header parsing is heavily dependent on list indexing. This is very confusing and lacks performance. A better approach should be analysed.

CalilQ commented on sharc/simulation.py:
```diff
@@ -195,15 +196,22 @@ def calculate_coupling_loss(self,
+                                                 shadowing=self.parameters.imt.shadowing,
+                                                 line_of_sight_prob=self.parameters.imt.line_of_sight_prob)
+            else:
+                #TODO Pass parameters correctly
```
This TODO can be removed

CalilQ committed:
Remove TODO message

CalilQ commented on sharc/propagation/propagation_factory.py:
```diff
@@ -41,6 +42,8 @@ def createPropagation(channel_model: str, param: Parameters) -> Propagation:
             return PropagationP619()
         elif channel_model == "INDOOR":
             return PropagationIndoor(param.indoor)
+        elif channel_model == "INPUT_FILES":
```
INPUT_FILE (singular) might be more appropriate

brunohcfaria commented on sharc/propagation/propagation_input_file.py:
```diff
@@ -0,0 +1,202 @@
+                    try:
+                        idx_list.append(head_list.index(param))
+                    except ValueError as err:
+                        if param in essential_params:
```
Shouldn't be this test outside the except clause?

brunohcfaria commented on tests/test_propagation_input_file.py:
```diff
@@ -0,0 +1,154 @@
+    
+    def setUp(self):
+        # Test 1
+        input_folder = "propagation_test_files\\test_1"
```
Change all file names and directories to OS independent.

brunohcfaria commented on sharc/propagation/propagation_input_file.py:
```diff
@@ -0,0 +1,202 @@
+        loss = np.zeros((len(bs_id),len(ue_position_x)))
+
+        # Loop through all the cells
+        for k in range(len(bs_id)):
```
Is `bs_id` a list of Base stations ids? If so, use `enumerate(bs_id)` to iterate over bs_id elements and indices.

brunohcfaria replied:
@CalilQ 
em vez de `for k in range(len(ds_id))`, que tal `for k, bs in enumerate(bs_id):`
Vc pode remover a linha 163: `bs = bs_id[k]`
é mais Pythonic...

brunohcfaria committed:
Fixed path name errors on propagation_input_file and test_propagation_input_file

brunohcfaria committed:
Fixed test_propagation_input_file tests

CalilQ committed:
Improve path loss file parsing

Parsing is not dependent on array indexing anymore

CalilQ committed:
Remove named tuple from path loss header

Structure is now a dictionary

CalilQ committed:
Improve loop through cells

Also fix some import errors

CalilQ committed:
Rename class for consistency

Instead of PropagationInputFile (singular), change to PropagationInputFiles (plural)

CalilQ reviewed (approved):
Code style and unit tests are ok. Result consistency and regression still need to be checked though.

brunohcfaria commented on sharc/propagation/propagation_input_file.py:
```diff
@@ -82,48 +82,37 @@ def __init__(self, input_folder: str):
                 # Parse other parameters
-                try: location = [float(x) for x in head_dict["LOCATION"]]
-                except KeyError: location = undefined
+                try: head["LOCATION"] = [float(x) 
```
@CalilQ O que vc acha de trocar o(s) try/except por isso aqui?
`
if 'LOCATION' not in head.keys():
    head['LOCATION'] = undefined
`
Geralmente a construção try/except é para algo excepcional.

CalilQ committed:
Substitute try/except with if/else

CalilQ committed:
Separate parameter files

Parameters for running input file simulations are in parameters_input_files.ini, while old sharing parameters are in parameters.ini
